src/TriangleNodesProvisioner.cpp

```diff
@@ -441,8 +444,116 @@ namespace blitzdg {
-        // WIP here...
+        firstIndex ii;
+        secondIndex jj;
+        thirdIndex kk;
+
+        index_matrix_type nodeIds(NumLocalPoints, NumElements);
+        
+        const index_matrix_type& Fmsk = *Fmask;
+        index_vector_type& vmM = *vmapM;
+        index_vector_type& vmP = *vmapP;
+        index_vector_type& mP = *mapP;
+
+        const index_vector_type& E2E = Mesh2D.get_EToE();
+        const index_vector_type& E2F = Mesh2D.get_EToF();
+        const index_vector_type& E2V = Mesh2D.get_Elements(); 
+
+        real_matrix_type xmat(NumLocalPoints, NumElements, ColumnMajorOrder());
+        real_matrix_type ymat(NumLocalPoints, NumElements, ColumnMajorOrder());
+
+        xmat = *xGrid; ymat = *yGrid;
+
+        real_vector_type x(NumElements*NumLocalPoints);
+        real_vector_type y(NumElements*NumLocalPoints);
+        reshapeMatTo1D(xmat, x.data(), false);
+        reshapeMatTo1D(ymat, y.data(), false);
+
+        // Assemble global volume node numbering.
+        nodeIds = ii + NumLocalPoints*jj;
+
+        Array<index_type, 3> vmapM3(NumFacePoints, NumFaces, NumElements);
+        Array<index_type, 3> vmapP3(NumFacePoints, NumFaces, NumElements);
+        Array<index_type, 3> mapP3(NumFacePoints, NumFaces, NumElements);
+
+        vmapM3 = 0*kk; vmapP3 = 0*kk; mapP3 = 0*kk;
+
+        for (index_type k=0; k < NumElements; ++k) {
+            for (index_type f=0; f < NumFaces; ++f) {
+                for (index_type n=0; n < NumFacePoints; ++n) {
+                    vmapM3(n,f,k) = nodeIds(Fmsk(n,f), k);
+                }
+            }
+        }
+
+        // Find index of face nodes with respect to volume node ordering.
+        for (index_type k=0; k <  NumElements; ++k) {
+            for (index_type f=0; f < NumFaces; ++f) {
+                for (index_type n=0; n < NumFacePoints; ++n) {
+                    // find neighbor.
+                    index_type k2 = E2E(NumFaces*k + f);
+                    index_type f2 = E2F(NumFaces*k + f);
+                    const real_vector_type Vert = Mesh2D.get_Vertices();                    
+
+                    index_type v1 = E2V(k*NumFaces + f); 
+                    index_type v2 = E2V(k*NumFaces + ((f+1) % NumFaces));
+
+                    // Compute reference length of edge.
+                    real_type vx1 = Vert(3*v1), vy1 = Vert(3*v1 + 1);
+                    real_type vx2 = Vert(3*v2), vy2 = Vert(3*v2 + 1);
+
+                    real_type refd = hypot(vx1-vx2, vy1-vy2);
+
+                    // find find volume node numbers of left and right nodes 
+                    index_type vidM = vmapM3(n,f,k);
+                    real_type x1 = x(vidM), y1 = y(vidM);
+                    
+                    for (index_type nP=0; nP < NumFacePoints; ++nP) {
+                        index_type vidP = vmapM3(nP,f2,k2);
+                        real_type x2 = x(vidP), y2 = y(vidP);
+
+                        if (hypot(x2-x1, y2-y1) < refd*NodeTol) {
```

Comment:
The statement tested by the if statement should probably be its own distance function.